sam_view.c

```diff
@@ -347,6 +421,45 @@ int main_samview(int argc, char *argv[])
                 ret = 1;
                 goto view_end;
             }
+            break;
+        case 'd':
+            if (strlen(optarg) < 4 || optarg[2] != ':') {
+                print_error_errno("view", "Invalid barcode option: \"%s\"", optarg);
+                ret = 1;
+                goto view_end;
+            }
+
+            {
+            char* tmp = strdup(optarg);
```

Comment:
You could do away with one `strdup `(and the need to check its return value) with something like:
```
            settings.barcode_tag = calloc(3, 1);
            if (!settings.barcode_tag) {
                print_error("view", "Could not allocate memory for tag: \"%s\"", optarg);
                ret = 1;
                goto view_end;
            }
            memcpy(settings.barcode_tag, optarg, 2);

            if (add_barcode_single("view", &settings, optarg+3) != 0) {
                ret = 1;
                goto view_end;
            } 
```

Reply:
You mean like this?
```diff
$ git diff sam_view.c
diff --git a/sam_view.c b/sam_view.c
index 3bf7dab..b099af6 100644
--- a/sam_view.c
+++ b/sam_view.c
@@ -429,16 +429,16 @@ int main_samview(int argc, char *argv[])
                 goto view_end;
             }
 
-            {
-            char* tmp = strdup(optarg);
-            tmp[2] = '\0';
-            settings.barcode_tag = strdup(tmp);
-            if (add_barcode_single("view", &settings, &tmp[3]) != 0) {
-                free(tmp);
+            if (!(settings.barcode_tag = calloc(3, 1))) {
+                print_error("view", "Could not allocate memory for tag: \"%s\"", optarg);
                 ret = 1;
                 goto view_end;
             }
-            free(tmp);
+            memcpy(settings.barcode_tag, optarg, 2);
+
+            if (add_barcode_single("view", &settings, optarg+3) != 0) {
+                ret = 1;
+                goto view_end;
             }
             break;
         case 'D':
```